A40-csds-support.md

```diff
@@ -0,0 +1,250 @@
+the interpreted structure in each stack, so there is no memory management logic
+change needed.
+
+#### Detail: Expose xDS Config As Is
+
+To date, gRPC supports the majority of xDS config but not all. Alternatively, we
+could only dump the config that gRPC understands. However, doing so will create
+a behavior difference between the gRPC and the control plane. It might be
+confusing to users that some fields are missing. Also, during the configuration
+dump, Envoy also dumps its entire configuration even if it doesn’t understand
+some of the configuration fields (think of fields for Envoy extensions). So, for
+correctness and simplicity, we should **cache and expose the xDS config the way
+gRPC receives them**.
+
+#### Detail: Node Matching
+
+The node matching mechanism is not designed for gRPC’s use case, but for
+querying the control plane. This doc proposes to ignore the Node matching query.
+
+
+## Alternatives
+
+### Solution: Config Dump By Attaching An HTTP Server
+
+Although it is the most straightforward way of implementing an admin interface,
+Java doesn’t have a good enough built-in web server and it is challenging to
+introduce yet another dependency into Core. Exposing application states via HTTP
+will be challenging from engineering perspective.
+
+### Solution: Config Dump via Channelz
+
+Though only a subset of xDS resources is applicable to a channel, xDS configs
+(listeners, routes, clusters, endpoints) are global resources to a gRPC
+application. One unsolved problem is that if we choose to use CSDS to expose xDS
+configs, **the users won’t have the ability to query the set of effective xDS
+configs for a channel**. The CSDS is not designed for finer granularity than
+individual processes.
+
+If we decided to support sharing XdsClient across multiple channels, the
+boundary of each channel’s xDS config will be blurry. Injecting xDS info into a
+channel tracing service won’t be the right direction in the long term.
+
+### Solution: Config Dump via File
+
+This approach saves the active xDS configuration onto a memory-based file system
+location, like `sysctl`, or Envoy Runtime. However, due to the lack of two-way
+communication and the complexity of the new machinery, this solution could cost
+more engineering resources but yield less functionality.
+
+
+## Implementation
+
+### CSDS Proto Updates
+
+* [envoy#13121]: the config synchronization status was defined from the xDS
+  management server point of view, this PR adds a set of ENUM to present the
+  config synchronization status from a client-side view.
+* [envoy#14689]: the CSDS focused on dumping the in-effective xDS configs, but
+  we can do better to improve gRPC’s debuggability. This PR adds fields to
+  config dump protos to allow CSDS to return information about the rejected
+  update.
+* [envoy#14900]: this PR adds two additional status to client configs, the
+  REQUESTED and the DOES_NOT_EXIST (read more).
```

Comment:
What's `(read more)`?

Reply:
Good catch! It's https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#knowing-when-a-requested-resource-does-not-exist. Link updated.